rust/connlib/tunnel/src/dns.rs

```diff
@@ -592,32 +646,42 @@ mod tests {
 
-        assert_eq!(match_domain(&domain("a.foo.com"), &resources), Some(0));
-        assert_eq!(match_domain(&domain("foo.com"), &resources), Some(0));
-        assert_eq!(match_domain(&domain("a.b.foo.com"), &resources), Some(0));
-        assert_eq!(match_domain(&domain("oo.com"), &resources), Some(1));
-        assert_eq!(match_domain(&domain("oo.xyz"), &resources), None);
+        assert_eq!(pattern.to_string(), "**.example.com")
     }
 
-    #[test]
-    fn question_mark_matching() {
-        let resources = HashMap::from([("?.bar.com".to_string(), 1)]);
+    #[test_case("**.example.com", "example.com"; "double star matches root domain")]
+    #[test_case("app.**.example.com", "app.bar.foo.example.com"; "double star matches multiple levels within domain")]
+    #[test_case("**.example.com", "foo.example.com"; "double star matches one level")]
+    #[test_case("**.example.com", "foo.bar.example.com"; "double star matches two levels")]
+    #[test_case("*.example.com", "foo.example.com"; "single star matches one level")]
+    #[test_case("*.example.com", "example.com"; "single star matches root domain")]
+    #[test_case("foo.*.example.com", "foo.bar.example.com"; "single star matches one domain within domain")]
+    #[test_case("app.*.*.example.com", "app.foo.bar.example.com"; "single star can appear on multiple levels")]
+    #[test_case("app.f??.example.com", "app.foo.example.com"; "question mark matches one letter")]
+    #[test_case("app.example.com", "app.example.com"; "matches literal domain")]
+    fn domain_pattern_matches(pattern: &str, domain: &str) {
+        let pattern = Pattern::new(pattern).unwrap();
+        let candidate = Candidate::from_str(domain).unwrap();
 
-        assert_eq!(match_domain(&domain("a.bar.com"), &resources), Some(1));
-        assert_eq!(match_domain(&domain("bar.com"), &resources), Some(1));
-        assert_eq!(match_domain(&domain("a.b.bar.com"), &resources), None);
+        let matches = pattern.matches(&candidate);
+
+        assert!(matches);
     }
 
-    #[test]
-    fn exact_matching() {
-        let resources = HashMap::from([("baz.com".to_string(), 2)]);
+    #[test_case("app.*.example.com", "app.foo.bar.example.com"; "single star does not match two level")]
+    #[test_case("app.*com", "app.foo.com"; "single star does not match dot")]
+    // #[test_case("app.**com", "app.foo.com"; "double star does not match dot")]
+    #[test_case("app?com", "app.com"; "question mark does not match dot")]
+    fn domain_pattern_does_not_match(pattern: &str, domain: &str) {
+        let pattern = Pattern::new(pattern).unwrap();
+        let candidate = Candidate::from_str(domain).unwrap();
+
+        let matches = pattern.matches(&candidate);
 
-        assert_eq!(match_domain(&domain("baz.com"), &resources), Some(2));
-        assert_eq!(match_domain(&domain("a.baz.com"), &resources), None);
-        assert_eq!(match_domain(&domain("a.b.baz.com"), &resources), None);
+        assert!(!matches);
```

Comment:
Good start here. Should we allow `**` and `*` only? I think we should at least require a TLD and a dot.

I may suggest a few more edge cases because I know folks will enter them.

```
#[test_case("*?*.example.com", "app.example.com"; "matches mix of * and ?")]
#[test_case("**.com", "app.example.com"; "matches up to TLD")]
#[test_case("app.**.web.**.example.com", "app.web.example.com"; "matches multiple recursive globs")]

Reply:
> #[test_case("**.com", "app.example.com"; "matches up to TLD")]

This one is redundant with line 658 I believe. The other two I added.